ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+
+The graph in our example is particularly simple::
+
+  <Graph name="graph1" type="activities">
+   <xgmml>
+    <graph wfid="2">
+     <node id="1">
+      <att>
+       <graph>
+        <att name="rootGraph" value="1"/>
+        <edge id="2_0" source="2" target="3"/>
+        <node id="2" label="Index Read&#10;&apos;names&apos;">
+         ... attributes for activity 2 ...
+        </node>
+        <node id="3" label="Output&#10;Result #1">
+         ... attributes for activity 3 ...
+        </node>
+       </graph>
+      </att>
+     </node>
+    </graph>
+   </xgmml>
+  </Graph>
+
+This graph contains a single subgraph (node id=2) which contains two activities - an index read activity and an output result
+activity.  These activities are linked by a single edge (id "2_0").  The details of the contents are covered in the section on
+executing graphs below.
+
+Generated activity helpers
+==========================
+
+Each activity has a corresponding class instance in the generated code, and a factory function for creating instances of that class::
+
+    struct cAc2 : public CThorIndexReadArg {
+        ... Implementation of the helper for activity #2 ...
+    };
+    extern "C" ECL_API IHThorArg * fAc2() { return new cAc2; }
+    
+    struct cAc3 : public CThorWorkUnitWriteArg {
+        ... Implementation of the helper for activity #3 ...
+    };
+    extern "C" ECL_API IHThorArg * fAc3() { return new cAc3; }
+
+The helper class for an activity implements the interface that is required for that particular kind.  (The interfaces are defined in
+rtl/include/eclhelper.hpp - many more details on that below.)
+
+Other
+=====
+The are other items that are logically associated with a workunit.  They may be stored in the workunit dll or in
+various places in dali (or cassandra), but they are all accessed through the IWorkUnit interface in
+common/workunit/workunit.hpp which hides the implementation details.  For instance information generated at runtime cannot
+by definition be included in the workunit dll.
+
+Options
+-------
+
+Options that are supplied to eclcc via the -f command line option, or the #option statement are included in the <Debug> section of the workunit xml::
+
+ <Debug>
+  <addtimingtoworkunit>0</addtimingtoworkunit>
+  <noterecordsizeingraph>1</noterecordsizeingraph>
+  <showmetaingraph>1</showmetaingraph>
+  <showrecordcountingraph>1</showrecordcountingraph>
+  <spanmultiplecpp>0</spanmultiplecpp>
+  <targetclustertype>hthor</targetclustertype>
+ </Debug>
+
+Note, the names of the options are case insensitive, and converted to lower case.
+
+Input Parameters
+----------------
+Many queries contain input parameters which modify their behaviour.  These
+correspond to STORED definitions in the ECL.  Our example contains a single string "searchName", so the
+workunit contains a single input parameter::
+
+ <Variables>
+  <Variable name="searchname">
+   <SchemaRaw xsi:type="SOAP-ENC:base64">
```

Comment:
Perhaps a brief explanation of why we use this encoding and/or just state that we *always* use this encoding.